### PR TITLE
Support test imperative basic in eager

### PR DESCRIPTION
**Pull request by JiabinYang**
### PR types

New features
### PR changes

Others
### Describe

In Eager mode, 13 unit tests in test_imperative_basic have been tuned by this PR. 
In order to, support these tests we add more logic to eager mode includes:

1. The kernel is called, and the output tensor is passed into the original pointer of the tensor on the python side, so that the tensor on the python side can be modified
2. Add logic to handle the automatic call timing of RetainGradForTensor
3. _varbase_creator, create_parameter, support the creation of EagerTensor in Eager mode, including EagerParamBase class
4. Dynamic graph, Parameter Initializer no longer uses append_op, but uses _C_ops.XXX
5. Support Initialzier FP16 tests in eager mode
6. Support all dygraph_utils in eager mode

### CODE REVIEW

File: paddle/fluid/eager/legacy/infer_var_type_context.h [hunk 1]
```diff
@@ -153,7 +153,8 @@ class TensorRuntimeInferVarTypeContext
 
   paddle::framework::proto::VarType::Type GetOutputType(
       const std::string& name, const int& index = 0) const override {
-    return paddle::framework::ToVarType(outputs_.at(name)[index]->Var().Type());
+    // TODO(jiabin): Support SelectedRows when we have it.
+    return paddle::framework::proto::VarType::LOD_TENSOR;
```
Comment (wanghuancoder) [inline paddle/fluid/eager/legacy/infer_var_type_context.h:157]: 这个问题我也遇到了，如果现在不解决，后面确定有解决方案吗？
如果我们在一个“没有解决方案”的方案上继续开发是有风险的。

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
LGTM for framework.py

**Review by phlrain — APPROVED** [review]

File: paddle/fluid/eager/legacy/infer_var_type_context.h (hunk 1 above)
Comment (JiabinYang) [reply paddle/fluid/eager/legacy/infer_var_type_context.h:157]: 这个其实是有解决方案的，但是需要和目前算子库项目有整体的适配，这个操作的本质是在Tensor没有初始化之前获取其Type，而这个Type目前尚未发现显式设置的地方，因此暂时使用硬编码的方式，后期已经留下了TODO，通过相应的算子库接口实现运行时判断

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**